src/main/subset.c

```diff
@@ -134,36 +145,52 @@ attribute_hidden SEXP ExtractSubset(SEXP x, SEXP indx, SEXP call)
     PROTECT(result = allocVector(mode, n));
     switch(mode) {
     case LGLSXP:
-	EXTRACT_SUBSET_LOOP(LOGICAL0(result)[i] = LOGICAL_ELT(x, ii),
-			    LOGICAL0(result)[i] = NA_INTEGER);
+	EXTRACT_SUBSET_LOOP(const int* px = (const int*) DATAPTR_OR_NULL(x),
+	                    LOGICAL0(result)[i] = px[ii],
+	                    LOGICAL0(result)[i] = LOGICAL_ELT(x, ii),
+			    LOGICAL0(result)[i] = NA_LOGICAL);
```

Comment:
I wonder if something along the following lines would work too, with a suitable implementation above:

```suggestion
	EXTRACT_SUBSET_LOOP(const int* px = (const int*) DATAPTR_OR_NULL(x),
	                    SET_LOGICAL_ELT, LOGICAL_ELT, NA_LOGICAL);
```

`SET_LOGICAL_ELT()` could be defined as a macro, like in `src/main/saveload.c` .